zcash_client_backend/src/fees/zip317.rs

```diff
@@ -67,132 +70,9 @@ impl ChangeStrategy for SingleOutputChangeStrategy {
-                (orchard.inputs().len(), orchard.outputs().len());
-            #[cfg(not(feature = "orchard"))]
-            let (orchard_inputs_len, orchard_outputs_len) = (0, 0);
-
-            let o_non_dust = orchard_inputs_len - orchard_dust.len();
-            let o_allowed_dust = (orchard_outputs_len + orchard_change).saturating_sub(o_non_dust);
-
-            let available_grace_inputs = self
-                .fee_rule
-                .grace_actions()
-                .saturating_sub(t_non_dust)
-                .saturating_sub(s_non_dust)
-                .saturating_sub(o_non_dust);
-
-            let mut t_disallowed_dust = transparent_dust.len().saturating_sub(t_allowed_dust);
-            let mut s_disallowed_dust = sapling_dust.len().saturating_sub(s_allowed_dust);
-            let mut o_disallowed_dust = orchard_dust.len().saturating_sub(o_allowed_dust);
-
-            if available_grace_inputs > 0 {
-                // If we have available grace inputs, allocate them first to transparent dust
-                // and then to Sapling dust followed by Orchard dust. The caller has provided
-                // inputs that it is willing to spend, so we don't need to consider privacy
-                // effects at this layer.
-                let t_grace_dust = available_grace_inputs.saturating_sub(t_disallowed_dust);
-                t_disallowed_dust = t_disallowed_dust.saturating_sub(t_grace_dust);
-
-                let s_grace_dust = available_grace_inputs
-                    .saturating_sub(t_grace_dust)
-                    .saturating_sub(s_disallowed_dust);
-                s_disallowed_dust = s_disallowed_dust.saturating_sub(s_grace_dust);
-
-                let o_grace_dust = available_grace_inputs
-                    .saturating_sub(t_grace_dust)
-                    .saturating_sub(s_grace_dust)
-                    .saturating_sub(o_disallowed_dust);
-                o_disallowed_dust = o_disallowed_dust.saturating_sub(o_grace_dust);
-            }
-
-            // Truncate the lists of inputs to be disregarded in input selection to just the
-            // disallowed lengths. This has the effect of prioritizing inputs for inclusion by the
-            // order of the original input slices, with the most preferred inputs first.
-            transparent_dust.reverse();
-            transparent_dust.truncate(t_disallowed_dust);
-            sapling_dust.reverse();
-            sapling_dust.truncate(s_disallowed_dust);
-            orchard_dust.reverse();
-            orchard_dust.truncate(o_disallowed_dust);
-
-            if !(transparent_dust.is_empty() && sapling_dust.is_empty() && orchard_dust.is_empty())
-            {
-                return Err(ChangeError::DustInputs {
-                    transparent: transparent_dust,
-                    sapling: sapling_dust,
-                    #[cfg(feature = "orchard")]
-                    orchard: orchard_dust,
-                });
-            }
```

Comment:
The reason why I deleted this code is that it was unreasonably hard to figure out whether this error could occur incorrectly. What it appears to have been trying to do is to check that there are no dust spends other than ones that will not increase the fee. But that *assumed* that the either caller was handling dust intelligently when deciding on inputs, *or* that the code in this method was filtering the inputs, neither of which are actually true.